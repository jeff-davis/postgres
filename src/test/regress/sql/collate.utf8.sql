/*
 * This test is for collations and character operations when using the
 * builtin provider with the C.UTF-8 locale.
 */

/* skip test if not UTF8 server encoding */
SELECT getdatabaseencoding() <> 'UTF8' AS skip_test \gset
\if :skip_test
\quit
\endif

SET client_encoding TO UTF8;

--
-- Test builtin PG_UNICODE_FAST locale.
--

CREATE TABLE builtin_test1 (
  t TEXT COLLATE PG_UNICODE_FAST
);
INSERT INTO builtin_test1 VALUES
  ('abc DEF'),
  ('ábc sßs ßss DÉF'),
  ('ǄxxǄ ǆxxǅ ǅxxǆ'),
  ('ȺȺȺ'),
  ('ⱥⱥⱥ'),
  ('ⱥȺ');

SELECT
    t, lower(t), initcap(t), upper(t),
    length(convert_to(t, 'UTF8')) AS t_bytes,
    length(convert_to(lower(t), 'UTF8')) AS lower_t_bytes,
    length(convert_to(initcap(t), 'UTF8')) AS initcap_t_bytes,
    length(convert_to(upper(t), 'UTF8')) AS upper_t_bytes
  FROM builtin_test1;

DROP TABLE builtin_test1;

-- test Final_Sigma
SELECT lower('ΑΣ' COLLATE PG_UNICODE_FAST); -- 0391 03A3
SELECT lower('ΑΣ0' COLLATE PG_UNICODE_FAST); -- 0391 03A3 0030
SELECT lower('ἈΣ̓' COLLATE PG_UNICODE_FAST); -- 0391 0343 03A3 0343
SELECT lower('ᾼΣͅ' COLLATE PG_UNICODE_FAST); -- 0391 0345 03A3 0345

-- test !Final_Sigma
SELECT lower('Σ' COLLATE PG_UNICODE_FAST); -- 03A3
SELECT lower('0Σ' COLLATE PG_UNICODE_FAST); -- 0030 03A3
SELECT lower('ΑΣΑ' COLLATE PG_UNICODE_FAST); -- 0391 03A3 0391
SELECT lower('ἈΣ̓Α' COLLATE PG_UNICODE_FAST); -- 0391 0343 03A3 0343 0391
SELECT lower('ᾼΣͅΑ' COLLATE PG_UNICODE_FAST); -- 0391 0345 03A3 0345 0391

-- properties

SELECT 'xyz' ~ '[[:alnum:]]' COLLATE PG_UNICODE_FAST;
SELECT 'xyz' !~ '[[:upper:]]' COLLATE PG_UNICODE_FAST;
SELECT '@' !~ '[[:alnum:]]' COLLATE PG_UNICODE_FAST;
SELECT '=' !~ '[[:punct:]]' COLLATE PG_UNICODE_FAST; -- symbols are not punctuation
SELECT 'a8a' ~ '[[:digit:]]' COLLATE PG_UNICODE_FAST;
SELECT '൧' ~ '\d' COLLATE PG_UNICODE_FAST;

-- case mapping

SELECT 'xYz' ~* 'XyZ' COLLATE PG_UNICODE_FAST;
SELECT 'xAb' ~* '[W-Y]' COLLATE PG_UNICODE_FAST;
SELECT 'xAb' !~* '[c-d]' COLLATE PG_UNICODE_FAST;
SELECT 'Δ' ~* '[α-λ]' COLLATE PG_UNICODE_FAST;
SELECT 'δ' ~* '[Γ-Λ]' COLLATE PG_UNICODE_FAST; -- same as above with cases reversed

--
-- Test builtin C.UTF-8 locale.
--

CREATE COLLATION BUILTIN_C_UTF8 ( provider = builtin, locale = 'C.UTF-8' );

CREATE TABLE builtin_test2 (
  t TEXT COLLATE BUILTIN_C_UTF8
);
INSERT INTO builtin_test2 VALUES
  ('abc DEF'),
  ('ábc sßs ßss DÉF'),
  ('ǄxxǄ ǆxxǅ ǅxxǆ'),
  ('ȺȺȺ'),
  ('ⱥⱥⱥ'),
  ('ⱥȺ');

SELECT
    t, lower(t), initcap(t), upper(t),
    length(convert_to(t, 'UTF8')) AS t_bytes,
    length(convert_to(lower(t), 'UTF8')) AS lower_t_bytes,
    length(convert_to(initcap(t), 'UTF8')) AS initcap_t_bytes,
    length(convert_to(upper(t), 'UTF8')) AS upper_t_bytes
  FROM builtin_test2;

DROP TABLE builtin_test2;

-- negative test: Final_Sigma not used for builtin locale C.UTF-8
SELECT lower('ΑΣ' COLLATE BUILTIN_C_UTF8);
SELECT lower('ΑͺΣͺ' COLLATE BUILTIN_C_UTF8);
SELECT lower('Α΄Σ΄' COLLATE BUILTIN_C_UTF8);

-- properties

SELECT 'xyz' ~ '[[:alnum:]]' COLLATE BUILTIN_C_UTF8;
SELECT 'xyz' !~ '[[:upper:]]' COLLATE BUILTIN_C_UTF8;
SELECT '@' !~ '[[:alnum:]]' COLLATE BUILTIN_C_UTF8;
SELECT '=' ~ '[[:punct:]]' COLLATE BUILTIN_C_UTF8; -- symbols are punctuation in posix
SELECT 'a8a' ~ '[[:digit:]]' COLLATE BUILTIN_C_UTF8;
SELECT '൧' !~ '\d' COLLATE BUILTIN_C_UTF8; -- only 0-9 considered digits in posix

-- case mapping

SELECT 'xYz' ~* 'XyZ' COLLATE BUILTIN_C_UTF8;
SELECT 'xAb' ~* '[W-Y]' COLLATE BUILTIN_C_UTF8;
SELECT 'xAb' !~* '[c-d]' COLLATE BUILTIN_C_UTF8;
SELECT 'Δ' ~* '[α-λ]' COLLATE BUILTIN_C_UTF8;
SELECT 'δ' ~* '[Γ-Λ]' COLLATE BUILTIN_C_UTF8; -- same as above with cases reversed

DROP COLLATION BUILTIN_C_UTF8;
