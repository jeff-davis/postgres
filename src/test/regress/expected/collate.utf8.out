/*
 * This test is for collations and character operations when using the
 * builtin provider with the C.UTF-8 locale.
 */
/* skip test if not UTF8 server encoding */
SELECT getdatabaseencoding() <> 'UTF8' AS skip_test \gset
\if :skip_test
\quit
\endif
SET client_encoding TO UTF8;
--
-- Test builtin PG_UNICODE_FAST locale.
--
CREATE TABLE builtin_test1 (
  t TEXT COLLATE PG_UNICODE_FAST
);
INSERT INTO builtin_test1 VALUES
  ('abc DEF'),
  ('ábc sßs ßss DÉF'),
  ('ǄxxǄ ǆxxǅ ǅxxǆ'),
  ('ȺȺȺ'),
  ('ⱥⱥⱥ'),
  ('ⱥȺ');
SELECT
    t, lower(t), initcap(t), upper(t),
    length(convert_to(t, 'UTF8')) AS t_bytes,
    length(convert_to(lower(t), 'UTF8')) AS lower_t_bytes,
    length(convert_to(initcap(t), 'UTF8')) AS initcap_t_bytes,
    length(convert_to(upper(t), 'UTF8')) AS upper_t_bytes
  FROM builtin_test1;
        t        |      lower      |     initcap      |       upper       | t_bytes | lower_t_bytes | initcap_t_bytes | upper_t_bytes 
-----------------+-----------------+------------------+-------------------+---------+---------------+-----------------+---------------
 abc DEF         | abc def         | Abc Def          | ABC DEF           |       7 |             7 |               7 |             7
 ábc sßs ßss DÉF | ábc sßs ßss déf | Ábc Sßs Ssss Déf | ÁBC SSSS SSSS DÉF |      19 |            19 |              19 |            19
 ǄxxǄ ǆxxǅ ǅxxǆ  | ǆxxǆ ǆxxǆ ǆxxǆ  | ǅxxǆ ǅxxǆ ǅxxǆ   | ǄXXǄ ǄXXǄ ǄXXǄ    |      20 |            20 |              20 |            20
 ȺȺȺ             | ⱥⱥⱥ             | Ⱥⱥⱥ              | ȺȺȺ               |       6 |             9 |               8 |             6
 ⱥⱥⱥ             | ⱥⱥⱥ             | Ⱥⱥⱥ              | ȺȺȺ               |       9 |             9 |               8 |             6
 ⱥȺ              | ⱥⱥ              | Ⱥⱥ               | ȺȺ                |       5 |             6 |               5 |             4
(6 rows)

DROP TABLE builtin_test1;
-- test Final_Sigma
SELECT lower('ΑΣ' COLLATE PG_UNICODE_FAST); -- 0391 03A3
 lower 
-------
 ας
(1 row)

SELECT lower('ΑΣ0' COLLATE PG_UNICODE_FAST); -- 0391 03A3 0030
 lower 
-------
 ας0
(1 row)

SELECT lower('ἈΣ̓' COLLATE PG_UNICODE_FAST); -- 0391 0343 03A3 0343
 lower 
-------
 ἀς̓
(1 row)

SELECT lower('ᾼΣͅ' COLLATE PG_UNICODE_FAST); -- 0391 0345 03A3 0345
 lower 
-------
 ᾳςͅ
(1 row)

-- test !Final_Sigma
SELECT lower('Σ' COLLATE PG_UNICODE_FAST); -- 03A3
 lower 
-------
 σ
(1 row)

SELECT lower('0Σ' COLLATE PG_UNICODE_FAST); -- 0030 03A3
 lower 
-------
 0σ
(1 row)

SELECT lower('ΑΣΑ' COLLATE PG_UNICODE_FAST); -- 0391 03A3 0391
 lower 
-------
 ασα
(1 row)

SELECT lower('ἈΣ̓Α' COLLATE PG_UNICODE_FAST); -- 0391 0343 03A3 0343 0391
 lower 
-------
 ἀσ̓α
(1 row)

SELECT lower('ᾼΣͅΑ' COLLATE PG_UNICODE_FAST); -- 0391 0345 03A3 0345 0391
 lower 
-------
 ᾳσͅα
(1 row)

-- properties
SELECT 'xyz' ~ '[[:alnum:]]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT 'xyz' !~ '[[:upper:]]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT '@' !~ '[[:alnum:]]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT '=' !~ '[[:punct:]]' COLLATE PG_UNICODE_FAST; -- symbols are not punctuation
 ?column? 
----------
 t
(1 row)

SELECT 'a8a' ~ '[[:digit:]]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT '൧' ~ '\d' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

-- case mapping
SELECT 'xYz' ~* 'XyZ' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT 'xAb' ~* '[W-Y]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT 'xAb' !~* '[c-d]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT 'Δ' ~* '[α-λ]' COLLATE PG_UNICODE_FAST;
 ?column? 
----------
 t
(1 row)

SELECT 'δ' ~* '[Γ-Λ]' COLLATE PG_UNICODE_FAST; -- same as above with cases reversed
 ?column? 
----------
 t
(1 row)

--
-- Test builtin C.UTF-8 locale.
--
CREATE COLLATION BUILTIN_C_UTF8 ( provider = builtin, locale = 'C.UTF-8' );
CREATE TABLE builtin_test2 (
  t TEXT COLLATE BUILTIN_C_UTF8
);
INSERT INTO builtin_test2 VALUES
  ('abc DEF'),
  ('ábc sßs ßss DÉF'),
  ('ǄxxǄ ǆxxǅ ǅxxǆ'),
  ('ȺȺȺ'),
  ('ⱥⱥⱥ'),
  ('ⱥȺ');
SELECT
    t, lower(t), initcap(t), upper(t),
    length(convert_to(t, 'UTF8')) AS t_bytes,
    length(convert_to(lower(t), 'UTF8')) AS lower_t_bytes,
    length(convert_to(initcap(t), 'UTF8')) AS initcap_t_bytes,
    length(convert_to(upper(t), 'UTF8')) AS upper_t_bytes
  FROM builtin_test2;
        t        |      lower      |     initcap     |      upper      | t_bytes | lower_t_bytes | initcap_t_bytes | upper_t_bytes 
-----------------+-----------------+-----------------+-----------------+---------+---------------+-----------------+---------------
 abc DEF         | abc def         | Abc Def         | ABC DEF         |       7 |             7 |               7 |             7
 ábc sßs ßss DÉF | ábc sßs ßss déf | Ábc Sßs ßss Déf | ÁBC SßS ßSS DÉF |      19 |            19 |              19 |            19
 ǄxxǄ ǆxxǅ ǅxxǆ  | ǆxxǆ ǆxxǆ ǆxxǆ  | Ǆxxǆ Ǆxxǆ Ǆxxǆ  | ǄXXǄ ǄXXǄ ǄXXǄ  |      20 |            20 |              20 |            20
 ȺȺȺ             | ⱥⱥⱥ             | Ⱥⱥⱥ             | ȺȺȺ             |       6 |             9 |               8 |             6
 ⱥⱥⱥ             | ⱥⱥⱥ             | Ⱥⱥⱥ             | ȺȺȺ             |       9 |             9 |               8 |             6
 ⱥȺ              | ⱥⱥ              | Ⱥⱥ              | ȺȺ              |       5 |             6 |               5 |             4
(6 rows)

DROP TABLE builtin_test2;
-- negative test: Final_Sigma not used for builtin locale C.UTF-8
SELECT lower('ΑΣ' COLLATE BUILTIN_C_UTF8);
 lower 
-------
 ασ
(1 row)

SELECT lower('ΑͺΣͺ' COLLATE BUILTIN_C_UTF8);
 lower 
-------
 αͺσͺ
(1 row)

SELECT lower('Α΄Σ΄' COLLATE BUILTIN_C_UTF8);
 lower 
-------
 α΄σ΄
(1 row)

-- properties
SELECT 'xyz' ~ '[[:alnum:]]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT 'xyz' !~ '[[:upper:]]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT '@' !~ '[[:alnum:]]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT '=' ~ '[[:punct:]]' COLLATE BUILTIN_C_UTF8; -- symbols are punctuation in posix
 ?column? 
----------
 t
(1 row)

SELECT 'a8a' ~ '[[:digit:]]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT '൧' !~ '\d' COLLATE BUILTIN_C_UTF8; -- only 0-9 considered digits in posix
 ?column? 
----------
 t
(1 row)

-- case mapping
SELECT 'xYz' ~* 'XyZ' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT 'xAb' ~* '[W-Y]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT 'xAb' !~* '[c-d]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT 'Δ' ~* '[α-λ]' COLLATE BUILTIN_C_UTF8;
 ?column? 
----------
 t
(1 row)

SELECT 'δ' ~* '[Γ-Λ]' COLLATE BUILTIN_C_UTF8; -- same as above with cases reversed
 ?column? 
----------
 t
(1 row)

DROP COLLATION BUILTIN_C_UTF8;
